### PR TITLE
[phishing-controller] Add `PhishingDetector` from `eth-phishing-detector`

### PR DESCRIPTION
**Pull request by mikesposito**
## Explanation

This PR adds the `PhishingDetector` class from `eth-phising-detector`, along with some utility functions into the `@metamask/phishing-controller` package, converting all to Typescript and Jest.

### `@metamask/phishing-detector`

- **ADDED**: Added `PhishingDetector` from `eth-phising-detector`
  - The class, utils and tests have been converted to typescript

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

**Comment by mikesposito** [comment]
Main files from `eth-phishing-detector` original package have been added in this commit 1775471f86163696226976dbbaf6c8dace6d0dfc for an easier diff

File: packages/phishing-controller/src/PhishingDetector.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+  version?: string;
+  result: boolean;
+  match?: string;
+  type: 'all' | 'fuzzy' | 'blocklist' | 'allowlist' | 'blacklist' | 'whitelist';
+};
+
+export class PhishingDetector {
+  #configs: PhishingDetectorConfiguration[];
+
+  #legacyConfig: boolean;
```
Comment (mikesposito) [inline packages/phishing-controller/src/PhishingDetector.ts:58]: These two were originally public in `eth-phishing-detector`, but looks like they are not used outside of this class - we should be able to prepend # to these without issues

File: packages/phishing-controller/src/PhishingDetector.ts [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+  | PhishingDetectorList[];
+
+export type PhishingDetectorConfiguration = {
+  name?: string;
+  version?: number | string;
+  allowlist: string[][];
+  blocklist: string[][];
+  fuzzylist: string[][];
+  tolerance: number;
+};
```
Comment (mikesposito) [inline packages/phishing-controller/src/PhishingDetector.ts:45]: This might be confusing, but the main difference between the options passed to the constructor, and the configs saved in `#configs` (`PhishingDetectorConfiguration`) is that domains are divided in parts, so each list becomes a matrix

File: packages/phishing-controller/src/PhishingDetector.ts [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+      return {
+        match: result.match,
+        result: result.result,
+        type: legacyType,
+      };
+    }
+    return result;
+  }
+
+  #check(domain: string): PhishingDetectorResult {
```
Comment (mikesposito) [inline packages/phishing-controller/src/PhishingDetector.ts:115]: This can most likely be merged into the public `check` method since all the public method does is handle the differences with the legacy result type, so that we don't have two methods with the same name without a clear responsibility distinction.

I'm unsure about doing it in this PR though, since there are many things to review already

File: packages/phishing-controller/src/PhishingDetector.ts [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+  fuzzylist: string[][];
+  tolerance: number;
+};
+
+export type PhishingDetectorResult = {
+  name?: string;
+  version?: string;
+  result: boolean;
+  match?: string;
+  type: 'all' | 'fuzzy' | 'blocklist' | 'allowlist' | 'blacklist' | 'whitelist';
```
Comment (mikesposito) [inline packages/phishing-controller/src/PhishingDetector.ts:52]: This is a union between the legacy response and the new one. 

There is probably a better way to do this, and it most likely involves writing two different types of responses (one for legacy, one for the new one) and overloading the `check` method - but as it depends on how the class was instantiated we might need a generic type at the class level to know which of the two types the `check` method will return

**Comment by mikesposito** [comment]
@NicholasEllul exactly! the plan is to leave the `eth-phishing-detect` package for the configuration only. We'll remove the PhishingDetector implementation from that package once we have it here

File: packages/phishing-controller/src/PhishingDetector.test.ts [hunk 5]
```diff
@@ -0,0 +1,979 @@
...
+          const { result, type, name } = detector.check('fuzzy-first.com');
+
+          expect(result).toBe(true);
+          expect(type).toBe('fuzzy');
+          expect(name).toBe('first');
+        },
+      );
+    });
+
+    it('blocks a "similar enough" domain when it is in the fuzzylist of the first config', async () => {
```
Comment (mcmire) [inline packages/phishing-controller/src/PhishingDetector.test.ts:373]: Nit: Curious if this would describe the behavior better:

``` suggestion
    it('blocks a domain that is similar enough (within a tolerance) to a domain in the fuzzylist of the first config', async () => {

File: packages/phishing-controller/src/PhishingDetector.test.ts [hunk 6]
```diff
@@ -0,0 +1,979 @@
...
+          const { result, type, name } = detector.check('fuzzy-firstab.com');
+
+          expect(result).toBe(true);
+          expect(type).toBe('fuzzy');
+          expect(name).toBe('first');
+        },
+      );
+    });
+
+    it('allows a domain when it is in the fuzzylist of the first config but not similar enough', async () => {
```
Comment (mcmire) [inline packages/phishing-controller/src/PhishingDetector.test.ts:403]: Similar as above:

``` suggestion
    it('allows a domain that is not similar enough to a domain in the fuzzylist of the first config', async () => {

File: packages/phishing-controller/src/PhishingDetector.test.ts [hunk 7]
```diff
@@ -0,0 +1,979 @@
...
+          const { result, type, name } = detector.check('fuzzy-second.com');
+
+          expect(result).toBe(true);
+          expect(type).toBe('fuzzy');
+          expect(name).toBe('second');
+        },
+      );
+    });
+
+    it('blocks a "similar enough" domain when it is in the fuzzylist of the second config', async () => {
```
Comment (mcmire) [inline packages/phishing-controller/src/PhishingDetector.test.ts:462]: Similar as above:

``` suggestion
    it('blocks a domain that is similar enough (within a tolerance) to a domain in the fuzzylist of the second config', async () => {

File: packages/phishing-controller/src/PhishingDetector.test.ts [hunk 8]
```diff
@@ -0,0 +1,979 @@
...
+          const { result, type, name } = detector.check('fuzzy-secondab.com');
+
+          expect(result).toBe(true);
+          expect(type).toBe('fuzzy');
+          expect(name).toBe('second');
+        },
+      );
+    });
+
+    it('allows a domain when it is in the fuzzylist of the second config but not similar enough', async () => {
```
Comment (mcmire) [inline packages/phishing-controller/src/PhishingDetector.test.ts:492]:
``` suggestion
    it('allows a domain that is not similar enough to a domain in the fuzzylist of the second config', async () => {

File: packages/phishing-controller/src/PhishingDetector.test.ts [hunk 9]
```diff
@@ -0,0 +1,979 @@
...
+          const { result, type, name } = detector.check('fuzzy-both.com');
+
+          expect(result).toBe(true);
+          expect(type).toBe('fuzzy');
+          expect(name).toBe('first');
+        },
+      );
+    });
+
+    it('blocks a domain when it is in the first blocklist and the second fuzzylist (from first)', async () => {
```
Comment (mcmire) [inline packages/phishing-controller/src/PhishingDetector.test.ts:551]: This is a bit confusing. How about:

``` suggestion
    it('blocks a domain when it is in the first blocklist, even if it is also matched by the second fuzzylist', async () => {

File: packages/phishing-controller/src/PhishingDetector.test.ts [hunk 10]
```diff
@@ -0,0 +1,979 @@
...
+          const { result, type, name } = detector.check('blocked-by-first.com');
+
+          expect(result).toBe(true);
+          expect(type).toBe('blocklist');
+          expect(name).toBe('first');
+        },
+      );
+    });
+
+    it('blocks a domain when it is in the first fuzzylist and the second blocklist (from first)', async () => {
```
Comment (mcmire) [inline packages/phishing-controller/src/PhishingDetector.test.ts:581]:
``` suggestion
    it('blocks a domain when it is matched by the first fuzzylist, even if it is also in the second blocklist', async () => {

File: packages/phishing-controller/src/PhishingDetector.test.ts [hunk 11]
```diff
@@ -0,0 +1,979 @@
...
+        async ({ detector }) => {
+          const { result, type } = detector.check('blocked.com.');
+
+          expect(result).toBe(true);
+          expect(type).toBe('blocklist');
+        },
+      );
+    });
+
+    describe('with legacy config', () => {
```
Comment (mcmire) [inline packages/phishing-controller/src/PhishingDetector.test.ts:916]: Similar as a previous comment. All of the above tests seem to pertain to the recommended config. What do you think about grouping them together in a separate `describe`? So we'd have two describes, one for recommended, one for legacy.

File: packages/phishing-controller/src/PhishingDetector.test.ts [hunk 12]
```diff
@@ -0,0 +1,979 @@
...
+        async ({ detector }) => {
+          const { result, type } = detector.check('default.com');
+
+          expect(result).toBe(false);
+          expect(type).toBe('all');
+        },
+      );
+    });
+
+    it('blocks a domain when it is in the blocklist of the first config', async () => {
```
Comment (mcmire) [inline packages/phishing-controller/src/PhishingDetector.test.ts:251]: It would be nice to have some tests which verify the behavior with only one config, namely, what happens if the domain is present in just the allowlist, or the blocklist, or the fuzzylist with adequate tolerance, or the fuzzylist with inadequate tolerance; or even what happens if the domain is present in two lists at once. We can leave this for a separate PR, but just leaving this as a note here.

File: packages/phishing-controller/src/PhishingDetector.test.ts [hunk 13]
```diff
@@ -0,0 +1,979 @@
...
+          const { result, type } = detector.check('blocked.com.');
+
+          expect(result).toBe(true);
+          expect(type).toBe('blocklist');
+        },
+      );
+    });
+
+    describe('with legacy config', () => {
+      it('changes the type to whitelist when the result is allowlist', async () => {
```
Comment (mcmire) [inline packages/phishing-controller/src/PhishingDetector.test.ts:917]: It appears (based on the names of these tests) that we are treating the legacy API as an anomaly instead of placing it on equal footing as the recommended API.

We don't need to test that the detector works with multiple config objects, as you can't do that with the legacy API of course, but what are your thoughts on testing what happens if the domain is matched via fuzzylist? Or not matched via the fuzzylist?

File: packages/phishing-controller/src/PhishingDetector.ts [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+
+export type FuzzyTolerance =
+  | {
+      tolerance?: number;
+      fuzzylist: string[];
+    }
+  | {
+      tolerance?: never;
+      fuzzylist?: never;
+    };
```
Comment (mcmire) [inline packages/phishing-controller/src/PhishingDetector.ts:32]: What was your intent with using `never`, to indicate that these shouldn't be provided or something else? Curiously, if I hover over this type, then TypeScript converts the `never`s to `undefined`:

``` typescript
export type FuzzyTolerance =
  | {
      tolerance?: number;
      fuzzylist: string[];
    }
  | {
      tolerance?: never;
      fuzzylist?: never;
    };
```

However, I'm curious whether the `never`s are needed at all. The `?` already means that these properties are optional, so would this work?

``` suggestion
export type FuzzyTolerance = {
  tolerance?: number;
  fuzzylist: string[];
};
```

File: packages/phishing-controller/src/utils.test.ts [hunk 15]
```diff
@@ -140,3 +148,110 @@ describe('applyDiffs', () => {
...
+    expect(() => validateConfig({ name: 123 })).toThrow(
+      "Invalid config parameter: 'name'",
+    );
+  });
+
+  it('throws an error if the config contains an invalid version', () => {
+    expect(() => validateConfig({ version: { foo: 'bar' } })).toThrow(
+      "Invalid config parameter: 'version'",
+    );
+  });
```
Comment (mcmire) [inline packages/phishing-controller/src/utils.test.ts:173]: What do you think about adding a test for the happy path?

File: packages/phishing-controller/src/utils.test.ts [hunk 16]
```diff
@@ -140,3 +148,110 @@ describe('applyDiffs', () => {
...
+
+describe('processConfigs', () => {
+  it('correctly converts a list of configs to a list of processed configs', () => {
+    const configs = [
+      {
+        allowlist: ['example.com'],
+        blocklist: ['sub.example.com'],
+        fuzzylist: ['fuzzy.example.com'],
+        tolerance: 2,
+      },
```
Comment (mcmire) [inline packages/phishing-controller/src/utils.test.ts:203]: What are your thoughts on testing with multiple configs rather than just one?

File: packages/phishing-controller/src/utils.test.ts [hunk 17]
```diff
@@ -140,3 +148,110 @@ describe('applyDiffs', () => {
...
+    ]);
+  });
+});
+
+describe('matchPartsAgainstList', () => {
+  it('matches a domain against a list of parts', () => {
+    const domainParts = ['com', 'example'];
+    const list = [
+      ['com', 'example', 'sub'],
+      ['com', 'example'],
```
Comment (mcmire) [inline packages/phishing-controller/src/utils.test.ts:241]: What if you include an entry in the list that has fewer parts than `domainParts`? It should _not_ get matched, but if we are testing that it should not match a domain with more parts, it seems we should also test that it shouldn't match one with fewer.

``` suggestion
      ['com'],
      ['com', 'example'],

**Review by mcmire — COMMENTED** [review]
Nice work! A lot of comments here, but most are minor.

File: packages/phishing-controller/src/utils.test.ts (hunk 16 above)
Comment (legobeat) [reply packages/phishing-controller/src/utils.test.ts:203]: More sounds better but also extending with further tests maybe better in follow-up as this ports existing  code from an already utilized package?

File: packages/phishing-controller/src/PhishingDetector.ts (hunk 14 above)
Comment (mikesposito) [reply packages/phishing-controller/src/PhishingDetector.ts:32]: Good point. My intent here was to communicate through types that tolerance is used only in the presence of a fuzzylist, and in case of a missing fuzzylist (but with tolerance) the detector throws an error.

So, either the user passes a fuzzylist and _maybe_ a tolerance, or neither of the two.

With this change it would work:
```suggestion
export type FuzzyTolerance =
  | {
      tolerance?: number;
      fuzzylist: string[];
    }
  | {
      tolerance: never;
      fuzzylist: never;
    };
 ```
 
 Let me know if this makes sense

File: packages/phishing-controller/src/utils.ts [hunk 18]
```diff
@@ -81,3 +88,136 @@ export const applyDiffs = (
...
+  allowlist?: string[];
+  blocklist?: string[];
+  fuzzylist?: string[];
+  tolerance?: number;
+}): PhishingDetectorConfiguration => ({
+  allowlist: processDomainList(override?.allowlist || []),
+  blocklist: processDomainList(override?.blocklist || []),
+  fuzzylist: processDomainList(override?.fuzzylist || []),
+  tolerance: override?.tolerance || DEFAULT_TOLERANCE,
+});
```
[the remark this answers is not in the document]
Comment (mikesposito) [reply packages/phishing-controller/src/utils.ts:172]: Yes, the idea was to give default values to each property, in case a consumer would override only one of them

File: packages/phishing-controller/src/utils.ts [hunk 19]
```diff
@@ -81,3 +88,136 @@ export const applyDiffs = (
...
+
+  if (
+    'version' in config &&
+    (!['number', 'string'].includes(typeof config.version) ||
+      config.version === '')
+  ) {
+    throw new Error("Invalid config parameter: 'version'");
+  }
+
+  return true;
```
[the remark this answers is not in the document]
Comment (mikesposito) [reply packages/phishing-controller/src/utils.ts]: Ah right! I refactored it in ebdd87e8ae8c407af8a0a1f8b7b273a586c02e98

I also discovered that this does not work with arrow functions, so I changed this to a standard one

File: packages/phishing-controller/src/PhishingDetector.test.ts [hunk 20]
```diff
@@ -0,0 +1,979 @@
...
+  WithPhishingDetectorCallback<ReturnValue>,
+];
+
+/**
+ * Build a phishing detector and run a callback with it.
+ *
+ * @param args - The phishing detector options and callback.
+ * @returns The return value of the callback.
+ */
+async function withPhishingDetector<ReturnValue>(
```
[the remark this answers is not in the document]
Comment (mikesposito) [reply packages/phishing-controller/src/PhishingDetector.test.ts:971]: I don't think we need a clean-up after each test run, but I added this because ideally if all other classes under test use this pattern then it should be more easily recognizable and readable by a developer reading it for the first time.

But I don't have a strong opinion, we can also keep things "simple" and instantiate manually the detector in each test case

File: packages/phishing-controller/src/utils.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (mcmire) [reply packages/phishing-controller/src/utils.ts:169]: > Then the defaults for the rest of the fields will not be used with this approach, I think, as they only get filled in when the entire argument is missing?

No, if you pass `{ allowlist: [] }`, then it would fill in `[]` for `blocklist` and `fuzzylist` and `DEFAULT_TOLERANCE` for `tolerance`.

File: packages/phishing-controller/src/PhishingDetector.ts (hunk 14 above)
Comment (mcmire) [reply packages/phishing-controller/src/PhishingDetector.ts:32]: Gotcha. Okay, that makes sense.

File: packages/phishing-controller/src/PhishingDetector.ts (hunk 4 above)
Comment (mikesposito) [reply packages/phishing-controller/src/PhishingDetector.ts:82]: I agree, the legacy config support makes the detector logic a bit overcomplicated since it just changes the way the list is returned, and a breaking change here to cut the support would make more sense. 

But since there are some changes over the original code due to the TS migration, I would keep this for a follow-up PR, where we can cut the legacy config support.

As for the server-side tooling to update the config (and the config itself), that will stay in the original repo

File: packages/phishing-controller/src/utils.test.ts (hunk 16 above)
Comment (mikesposito) [reply packages/phishing-controller/src/utils.test.ts:203]: I'm also more inclined to leave improvements over the original code / tests for follow-up PRs, but definitely something we should do

File: packages/phishing-controller/src/PhishingDetector.test.ts (hunk 13 above)
Comment (mikesposito) [reply packages/phishing-controller/src/PhishingDetector.test.ts:917]: These test cases are meant to check if `allowlist` and `blocklist` are projected into `whitelist` and `blacklist` for legacy configs, but this is not the case with fuzzylists, which should always be reported as `fuzzy` also with legacy configs. 

We can add a test case like this anyway though:

```typescript
      it('uses the type `fuzzy` when the result is in fuzzylist', async () => {
        await withPhishingDetector(
          {
            blacklist: [],
            fuzzylist: ['fuzzy.com'],
            tolerance: 2,
            whitelist: [],
          },
          async ({ detector }) => {
            const { type, result } = detector.check('fupzy.com');

            expect(type).toBe('fuzzy');
            expect(result).toBe(true);
          },
        );
      });
```

File: packages/phishing-controller/src/PhishingDetector.test.ts [hunk 21]
```diff
@@ -0,0 +1,979 @@
...
+                version: mockInvalidVersion,
+              },
+            ],
+            async () => null,
+          ),
+        ).rejects.toThrow("Invalid config parameter: 'version'");
+      });
+    });
+
+    describe('with legacy config', () => {
```
[the remark this answers is not in the document]
Comment (mikesposito) [reply packages/phishing-controller/src/PhishingDetector.test.ts:160]: Good point, refactored in [38807df](https://github.com/MetaMask/core/pull/4137/commits/38807df51f5e6e16a083e5398de7512a3a07d7f3)

File: packages/phishing-controller/src/PhishingDetector.test.ts (hunk 11 above)
Comment (mikesposito) [reply packages/phishing-controller/src/PhishingDetector.test.ts:930]: Refactored in 38807df51f5e6e16a083e5398de7512a3a07d7f3

File: packages/phishing-controller/src/PhishingDetector.ts (hunk 4 above)
Comment (legobeat) [reply packages/phishing-controller/src/PhishingDetector.ts:82]: TIL! So the answer is actually "yes", but fully agreed such refactors are best done separately.

> This makes me think: If the check method returns different results depending on how the PhishingDetector is instantiated, should we have two PhishingDetectors? Almost like a PhishingDetectorV1 and PhishingDetectorV2, or a LegacyPhishingDetector and PhishingDetector. We can do that in a different PR of course, but just some food for thought.

@mcmire So due to the code in this repo no longer being concerned with the serving side, sounds like the door is even open for fully removing what have been `PhishingDetectorV1`!

Comment (mcmire) [reply packages/phishing-controller/src/PhishingDetector.ts:82]: Okay cool, that's good to know! I'm okay with the changes here and with removing V1 in another PR.

File: packages/phishing-controller/src/utils.test.ts (hunk 16 above)
Comment (mcmire) [reply packages/phishing-controller/src/utils.test.ts:203]: Okay no problem.

File: packages/phishing-controller/src/utils.test.ts (hunk 17 above)
Comment (mcmire) [reply packages/phishing-controller/src/utils.test.ts:241]: Hmm okay. It sounds like we may be testing this advertently. If that is the case, then I'm okay with making the tests more robust in another PR.

File: packages/phishing-controller/src/PhishingDetector.test.ts (hunk 13 above)
Comment (mcmire) [reply packages/phishing-controller/src/PhishingDetector.test.ts:931]: Ah I see, makes sense. That sounds like a good test to add.

Comment (mikesposito) [reply packages/phishing-controller/src/PhishingDetector.test.ts:931]: Added in [015fbd0](https://github.com/MetaMask/core/pull/4137/commits/015fbd0724255041953580473a490ce182b20945)

File: packages/phishing-controller/src/utils.test.ts (hunk 15 above)
Comment (mikesposito) [reply packages/phishing-controller/src/utils.test.ts:184]: Good idea! Added in [bf8fe91](https://github.com/MetaMask/core/pull/4137/commits/bf8fe913af9bd30e2b001e3d18ecd5f2a0d088f4)

**Review by mcmire — APPROVED** [review]
LGTM!
